FIP-0118: specify f02 mechanics: pull settlement, queue, validation (#1272)
filecoin-project/FIPs: FIPS/fip-0118.md

            flat over the segment (w2 from Q2 onward, between gate
            steps)

        3.  **`t_start`**: the epoch the segment starts from

        4.  **floor / cap**: clamp bounds

    3.  **Distribution**: the stream's recipients and how the stream
        is split among them. There are two possible values. `IMPLICIT`:
        nothing stored; f02 resolves the recipient from protocol
        state. The consensus stream is `IMPLICIT`, its recipient being
        the block winner chosen by leader election each epoch.
        `EXPLICIT`: the stream's stored recipient state, updated only by
        the stream's designated **writer** (the SRA for the service
        stream; never a payee). An `EXPLICIT` distribution holds the
        writer's address and three wallet-keyed tables:

        1.  **shares**: the current wallet-to-share map, at most
            `MAX_RECIPIENTS` entries, written via `SetShares`;

        2.  **payable**: per-wallet amounts carried over from closed
            periods, not yet claimed;

        3.  **`claimed_period`**: per-wallet amounts already claimed
            against the current period's accrual.

>         A future stream paying one fixed wallet needs no separate

Reviewer: "A future stream paying one fixed wallet needs no separate form" should be "A future stream paying any fixed set of wallets needs no separate form", no?

>         A future stream paying one fixed wallet needs no separate

Reviewer: ```suggestion
        A future stream paying any fixed set of wallets needs no separate
```

    weight-times-reward multiply promotes to a big integer and divides
    with floor. Slope quantization over the 9-quarter ramp is ~1e-10 of
    a percentage point at this scale, and the clamp endpoints are
    stored exact, so w1 comes to rest at exactly its 50% floor. Token
    amounts are attoFIL big integers, serialized as f02's existing
    fields are.

##### 2.4.2 State

f02's state is an 11-field tuple rewritten every block. The rule for
the changes: values that mutate per block stay inline in the root
tuple (`TokenAmount` counters and an epoch, both small on the wire);
everything else lives behind one CID, written only at governance,
quarterly or claim cadence.

| Field | Change |
|---|---|
| `total_minted_reward` (position 9) | Renamed from `total_storage_power_reward`, position kept. Accrues the full block reward, all streams (Section 2.5). |
| `simple_total`, `baseline_total` (positions 10, 11) | Deleted: retired to the code constants below, as the existing state comment anticipates ("can be moved from state back into a code constant in a subsequent upgrade"). |
| `total_burn_minted` | New counter: cumulative w0 residual, accrued at award. |

> | `total_service_minted` | New counter: cumulative accrual across all `EXPLICIT` streams. |

Reviewer: ```suggestion
| `total_explicit_minted` | New counter: cumulative accrual across all `EXPLICIT` streams. |
```

> | `service_accrued[id]` | New, one per `EXPLICIT` stream: the current period's gross accrual. |

Reviewer: ```suggestion
| `accrued[id]` | New, one per `EXPLICIT` stream: the current period's gross accrual. |
```

integers are byte strings: one sign byte (00 non-negative, 01
negative) then the big-endian magnitude, empty for zero. `Address`
is Filecoin address bytes, always ID form (protocol byte 00 then the
LEB128 id), normalized at every write. `Option<T>` is CBOR null or `T`.
Keyed arrays sort ascending by key with unique keys; `pending_writes`
alone is ordered by effective epoch, equal epochs retaining queue
position (the apply tiebreak). `Spacetime` is a big integer;
`FilterEstimate`
is a pair of big integers, unchanged.

```rust
pub struct State {                          // positions 1-8 unchanged
    cumsum_baseline: Spacetime,             //  1
    cumsum_realized: Spacetime,             //  2
    effective_network_time: ChainEpoch,     //  3
    effective_baseline_power: StoragePower, //  4
    this_epoch_reward: TokenAmount,         //  5
    this_epoch_reward_smoothed: FilterEstimate, // 6
    this_epoch_baseline_power: StoragePower, // 7
    epoch: ChainEpoch,                      //  8
    total_minted_reward: TokenAmount,       //  9  renamed
    total_burn_minted: TokenAmount,         // 10  new
    total_service_minted: TokenAmount,      // 11  new

>     service_accrued: Vec<StreamAccrual>,    // 12  new; ascending id

Reviewer: ```suggestion
    total_explicit_minted: TokenAmount,      // 11  new
    accrued: Vec<StreamAccrual>,    // 12  new; ascending id
```

    RegisterStream = 2,     // payload: [ WeightRecord, Option<DistributionInit> ]
    RemoveStream = 3,       // payload: [] (empty tuple)
    SetDistribution = 4,    // payload: [ Address ]
}
```

A pending write is a deferred method call, executed verbatim at
`effective_epoch` (Section 2.4.7). Each `EXPLICIT` stream owns its own
recipient tables, so the same wallet may appear in more than one
stream without collision, and a `Claim` rewrites only its own stream's
rows.

##### 2.4.3 The award path

Only `AwardBlockReward`'s internals change; its number, signature and
callers do not.

```
AwardBlockReward(miner, penalty, gas_reward, win_count):
    if epoch >= next_transition_epoch: apply due queued writes (Section 2.4.7)
    BR = this_epoch_reward * win_count / 5      // minted reward, as today
    paid = 0
    for each active stream s:                   // in list order
        portion = floor(ComputeWeight(s.weight_record, epoch) * BR)
        if s.distribution is IMPLICIT:          // consensus stream
            miner_reward = portion
        else:                                   // EXPLICIT: accrue
            service_accrued[s.id] += portion

>             total_service_minted += portion

Reviewer: ```suggestion
            accrued[s.id] += portion
            total_explicit_minted += portion
```

makes conservation exact to the atto, where flooring each weight
independently would not. Gas rewards remain wholly with the winning
miner. Nothing on this path can fail: the only send is to f099.

##### 2.4.4 SetShares and the period fold

A **period** in f02 is the interval between two `SetShares` calls on a
stream: f02 knows nothing of quarters and imposes no cadence, so the
writer may call `SetShares` at any time. The quarterly rhythm is SRA
discipline: under this FIP the SRA's only path to `SetShares` is
`SubmitShares`, once per quarter after the verification window (Section
3.2), and registry changes (admission, removal, freeze) take payment
effect at the next such submission. Installing a new map first
closes the current period under the old one, and that fold is what
makes immediate binding safe: earnings materialize under the shares
in force while they accrued, so a share change is strictly
prospective.

```
SetShares(id, new_map):     // designated writer only; never queued
    require caller == the stream's designated writer
    require Σ new_map shares == DENOM, every share positive
    resolve each recipient to an ID address, rejecting on failure

>     pool = service_accrued[id]

Reviewer: ```suggestion
    pool = accrued[id]
```

    for each (wallet, share) in the OLD map:
        earned = floor(share * pool)
        payable[wallet] += earned - claimed_period[wallet]
    residue = pool - Σ earned               // rounding dust only
    send(f099, residue)                     // burned; neither counter moves

>     service_accrued[id] = 0; clear claimed_period

Reviewer: ```suggestion
    accrued[id] = 0; clear claimed_period
```

    install new_map
```

A wallet dropped from the new map keeps its payable balance until it
claims. The residue is rounding dust only, because shares sum to
exactly one, and it burns without touching either counter:

> `total_service_minted` already counted it at accrual, so the

Reviewer: ```suggestion
`total_explicit_minted` already counted it at accrual, so the
```

decomposition attributes dust to the service stream and the derived
miner residual stays exact; f099's receipts exceed `total_burn_minted`
by cumulative dust. A recipient that does not resolve to an ID address
rejects the whole call: a backstop against typos and stranded funds,
with the SRA validating recipients at registration.

##### 2.4.5 Claim

`Claim(id, wallets[]) -> amounts[]` is permissionless and batched. Each
wallet's entitlement is its live portion of the current period,

> `floor(share * service_accrued[id])` minus what it has already claimed

Reviewer: ```suggestion
`floor(share * accrued[id])` minus what it has already claimed
```

constants of Section 2.4.2 on every network), zeroes the new
counters, and sets `swa_timelock_epochs`: the only place it is ever
written.

#### 2.5 Supply accounting

Circulating supply is consensus-relevant (`FilMined` feeds the initial
pledge input) and Lotus, Forest and Venus each compute it
independently, so its inputs should not change. The split changes who
receives issuance, not how much, so the total keeps its slot:

  - **`total_minted_reward`** (position 9, renamed from
    `total_storage_power_reward`) accrues the full block reward, all
    streams. The rename is source-level only: tuple serialization is
    positional, so nothing changes on the wire for state readers.
    **`FilMined` keeps reading position 9 unchanged**: no
    implementation changes its circulating-supply logic.

  - **`total_burn_minted`** (the w0 residual) and
    **`total_service_minted`** (gross service accrual) are stored
    counters, otherwise hard to recover independently: each needs the
    full weight and reward history. Fold dust burns from the service
    side and is attributed there (Section 2.4.4), keeping the residual
    exact.

  - What miners have received is the derived residual

>     `M = total_minted_reward − total_burn_minted − total_service_minted`;

Reviewer: ```suggestion
    **`total_explicit_minted`** (gross service accrual) are stored
    counters, otherwise hard to recover independently: each needs the
    full weight and reward history. Fold dust burns from the service
    side and is attributed there (Section 2.4.4), keeping the residual
    exact.

  - What miners have received is the derived residual
    `M = total_minted_reward − total_burn_minted − total_explicit_minted`;
```

    it is not stored. Tools that treated
    position 9 as miner income must move to this residual (see
    Backwards Compatibility).

  - Minting is counted at accrual: each award adds
    `BR = miner + service + burn` exactly and bumps the three totals by
    their parts, so the decomposition holds by construction. The burn
    is minted and sent to f099: counted in `FilMined`, captured by
    `FilBurnt`, net zero on circulating supply.

The claim state carries no supply duty: the counters account for FIL
ever minted, while the claim state accounts only for the slice still
inside f02. At every epoch f02's balance covers

> `Σ (service_accrued − Σ claimed_period + Σ payable)` over streams and

Reviewer: ```suggestion
`Σ (accrued − Σ claimed_period + Σ payable)` over streams and
```

bootstrap: the clamp does it. The move from 5% to `W2_BASE` is therefore
ungated: no gate check runs for Q1 (`last_checked_quarter` initializes
to 1), and the first `QuarterlyGateCheck` evaluates quarter 2's bound
volume, after Q2's verification window has closed.
From then on, w2 is a step function: when a
gate passes, `QuarterlyGateCheck` replaces the record with a constant
one at the new level,
`{ v_start: new_w2, slope: 0, floor = cap = new_w2 }`, via
`StepWeightRecords` under `SWA_TIMELOCK`; when a gate fails, the
record is untouched and w2 holds. Scheduled burn resumes
automatically from the Q1 boundary: w1 keeps ramping down while w2
holds at its gate-set level, and the residual is burned.

**`QuarterlyGateCheck`**, callable once per quarter, in quarter order,
the first check evaluating quarter 2:

```
// SWA state: steps = gate steps executed so far, indexing the target
// schedule directly; last_checked_quarter = the last quarter consumed
// by a check, initialized to 1 (the ungated bootstrap quarter).
func QuarterlyGateCheck():
    // Sequential once-per-quarter latch: quarters are checked in
    // order and each quarter's bound volume is evaluated exactly once

>     Q = last_checked_quarter + 1

Reviewer 2: I'm likely to track nextCheckedQuarter instead of lastCheckedQuarter in the SWA.
